lib/fs.js

```diff
@@ -415,6 +415,42 @@ function tryToString(buf, encoding, callback) {
   callback(e, buf);
 }
 
+function tryStatSync(fd, isUserFd) {
+  var threw = true;
+  var st;
+  try {
+    st = fs.fstatSync(fd);
+    threw = false;
+  } finally {
+    if (threw && !isUserFd) fs.closeSync(fd);
+  }
+  return st;
+}
+
+function tryCreateBuffer(size, fd, isUserFd) {
+  var threw = true;
+  var buffer;
+  try {
+    buffer = new Buffer(size);
+    threw = false;
+  } finally {
+    if (threw && !isUserFd) fs.closeSync(fd);
+  }
+  return buffer;
+}
+
+function tryReadSync(fd, isUserFd, buffer, pos, len) {
+  var threw = true;
+  var bytesRead;
+  try {
+    bytesRead = fs.readSync(fd, buffer, pos, len);
+    threw = false;
+  } finally {
+    if (threw && !isUserFd) fs.closeSync(fd);
+  }
+  return bytesRead;
+}
+
```

Comment:
Seeing these I can't help but wonder if there's value in having a more generic utility `attempt(fn,fin)` mechanism... e.g. 

```
const attempt = require('internal/util').attempt;
attempt(() => doSomething(), (threw) => doCleanup())
```

where `attempt()` == 

```
function attempt (fn, fin) {
  var threw = true;
  try {
    fn();
    threw = false;
  } finally {
    fin(threw);
  }
}
```

Likely not worth the additional complexity and perf cost tho.